cpp/src/utilities/stream_pool.cpp

```diff
@@ -147,11 +147,12 @@ struct cuda_event {
  */
 cudaEvent_t event_for_thread()
 {
-  thread_local std::vector<std::unique_ptr<cuda_event>> thread_events(get_num_cuda_devices());
+  // The program may crash if this function is called from the main thread and user application
+  // subsequently calls cudaDeviceReset().
+  // As a workaround, here we intentionally disable RAII and leak cudaEvent_t.
+  thread_local std::vector<cuda_event*> thread_events(get_num_cuda_devices());
   auto const device_id = get_current_cuda_device();
-  if (not thread_events[device_id.value()]) {
-    thread_events[device_id.value()] = std::make_unique<cuda_event>();
-  }
+  if (not thread_events[device_id.value()]) { thread_events[device_id.value()] = new cuda_event(); }
```

Comment:
explicitly delete the copy constructor and define the move constructor for cuda_event, otherwise there'll be a dangling handle and double-free in the future.

Reply:
Good catch!
Technically not related to the segfault fix, but it would be good to fix the potential source of bugs sooner rather than later :)